### PR TITLE
The mnist program with new fictive API

### PR DESCRIPTION
**Pull request by luotao1**
在 @reyoung 之前设计的一份[api_trainer.py](https://github.com/reyoung/Paddle/blob/9360a1fc15dd6ddbf220416be2161381b30feed9/demo/mnist/api_train.py)上修改了几个部分：
1. batch_size放到了optimizer地方，原来settings部分两者是放在一起的。
2. train_data和test_data用一个函数来获得，因为两者处理逻辑很类似。
3. 去掉多余的start和finish函数，也去掉原来的make_evaluator，希望把model.train的返回值直接作为evaluator输出。
4. 增加plot_cost函数，可以将train_cost和test_cost画在一张图上。

### CODE REVIEW

File: recognize_digits/example.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+Current implementation uses Raw SWIG, which means the API call is directly \
+passed to C++ side of Paddle.
+
+The user api could be simpler and carefully designed.
+"""
+
+import paddle.v2 as paddle
+
+from mnist_util import read_from_mnist
```
Comment (wangkuiyi) [inline recognize_digits/example.py]: read_from_mnist ==> read_mnist_data ?

File: recognize_digits/example.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+
+def main():
+    paddle.raw.initPaddle("-use_gpu=false",
+                          "-trainer_count=4")  # use 4 cpu cores
+
+    optimizer = paddle.optimizer.Optimizer(
```
Comment (wangkuiyi) [inline recognize_digits/example.py]: optimizer = paddle.optimizer.Adam(...) ?

File: recognize_digits/example.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        learning_method=paddle.optimizer.AdamOptimizer(),
+        learning_rate=1e-4,
+        model_average=paddle.optimizer.ModelAverage(average_window=0.5),
+        regularization=paddle.optimizer.L2Regularization(rate=0.5),
+        batch_size=128)  # 原来settings包含了batch_size
+
+    # define network
+    imgs = paddle.layers.data_layer(name='pixel', size=784)
```
Comment (wangkuiyi) [inline recognize_digits/example.py]: imgs = paddle.layer.data(...) ?

File: recognize_digits/example.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        input=inference, label=paddle.layers.data_layer(
+            name='label', size=10))
+
+    model = paddle.model.Model(layers=[cost], optimizer=optimizer)
+
+    model.rand_parameter()
+
+    # 希望把train_data和test_data写在一块，因为两者的处理逻辑很相似
+    # 去掉原来api设计中的model变量，这部分和model不相关
+    train_data, test_data = paddle.data.create_data_pool(
```
Comment (wangkuiyi) [inline recognize_digits/example.py]: pool 是个什么意思呢？是为了做data shuffling来打乱minibatch的顺序吗？如果只是为了这个，应该有一个更明确的名字，叫做 data_shuffler？

```
training_data = paddle.data.shuffler(read_mnist_data, './data/raw_data/train')
testing_data = paddel.data.shuffler(read_mnist_data, './data/raw_data/test')
```

File: recognize_digits/example.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        file_reader=read_from_mnist,
+        file_list=['./data/raw_data/train', './data/raw_data/test'],
+        shuffle=True)
+
+    # Training process.
+    model.start()
```
Comment (wangkuiyi) [inline recognize_digits/example.py]: 这个 start 函数要做什么呢？

另外，如果是 `trainer.start()` 那么大家可以顾名思义地理解start开始的动作是“training”；如果是 `tester.start()` 那么大家顾名思义理解理解start的是"testing"。可是 start model 是什么意思呢？是start testing using model还是start training a model还是。。。？看上去是有歧义的。

File: recognize_digits/example.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+    # 去掉多余的start和finish函数，也去掉原来的make_evaluator
+    # 能否把返回值直接作为evaluator输出呢？
+    for pass_id in xrange(2):
```
Comment (wangkuiyi) [inline recognize_digits/example.py]: 考虑到online learning的时候，数据流式没有终结的，我们无法定义pass这样的概念。我建议这里的双重循环改成单层： https://github.com/PaddlePaddle/book/pull/68#discussion_r99907558

File: recognize_digits/example.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+            print "Pass=%d, batch=%d" % (pass_id, batch_id), batch_evaluator
+
+        for _, data_batch in enumerate(test_data):
+            test_evaluator = model.test(data_batch)
+        print "TEST Pass=%d" % pass_id, test_evaluator
+
+        # 将训练和测试的cost画在一张图上
+        plot_cost(batch_evaluator.cost, test_evaluator.cost, pass_id)
+
+    model.finish()
```
Comment (wangkuiyi) [inline recognize_digits/example.py]: 这个model.finish是不是和model.start一样，其实是不需要的？

File: recognize_digits/example.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        shuffle=True)
+
+    # Training process.
+    model.start()
+
+    # 去掉多余的start和finish函数，也去掉原来的make_evaluator
+    # 能否把返回值直接作为evaluator输出呢？
+    for pass_id in xrange(2):
+        for batch_id, data_batch in enumerate(train_data):
+            batch_evaluator = model.train(data_batch)
```
Comment (wangkuiyi) [inline recognize_digits/example.py]: train这个动作包括了整个training loop。这里根据上下文，想要做的是update the model using a given minibatch，所以应该是下面这样吧？

```
batch_evaluator = model.update(minibatch)
```

Comment (helinwang) [inline recognize_digits/example.py]: 感觉需要一个方法让model.train知道data_batch里面哪些数据对应`imgs = paddle.layers.data_layer(name='pixel', size=784)`，哪些数据对应`paddle.layers.data_layer(name='label', size=10)`
